spec.md

```diff
@@ -0,0 +1,954 @@
+
+The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
+This string MAY be surfaced by CMI-Client to end users.
+
+#### `ListMachines`
+
+A Plugin MUST implement this RPC call if it has `LIST_MACHINES` capability.
+The Plugin SHALL return the information about all the machines associated with the `ProviderSpec`.
+The CMI-Client SHALL NOT expect a consistent "view" of all machines when paging through the machines list via multiple calls to `ListMachines`.
+
+- If the Plugin succeeded in returning a list of `machines-names` with their corresponding `machine-IDs`, then return `0 OK`.
+- The `ListMachineResponse` contains a map of `MachineList` whose
+    - Key is expected to contain the `MachineID` &
+    - Value is expected to contain the `MachineName` corresponding to it's kubernetes machine CR object
+- The plugin can OPTIONALY make use of the secrets supplied in the `Secrets` map in the `ListMachinesRequest` to communicate with the provider.
+
+```protobuf
+message ListMachinesRequest {
+    // ProviderSpec is the configuration needed to list machines.
+    // Driver should parse this raw data into pre-defined spec in their respective projects.
+    // This field is REQUIRED.
+    bytes ProviderSpec = 1;
+
+    // Secrets is the map containing necessary credentials for cloud-provider to list the machines.
+    // This field is OPTIONAL.
+    map<string, bytes> Secrets = 2 [(cmi_secret) = true];
+}
+
+message ListMachinesResponse {
+    // MachineList is the map of list of machines. Format for the map should be map<MachineID, MachineName>.
+    // This field is REQUIRED.
+    map<string, string> MachineList = 1;
+}
+```
+
+##### ListMachines Errors
+
+If the plugin is unable to complete the ListMachines call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CMI-Client MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
+|-----------|-----------|-------------|-------------------|------------|
+| 0 OK | Successful | The call for listing all VMs associated with `ProviderSpec` was successful. |  | N |
+| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
+| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
+| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `ProviderSpec` and make sure that all required fields are present in their desired value format. Exact issue to be given in `.message` | Update `ProviderSpec` to fix issues. | N |
+| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
```

Comment:
Can timeout happen because of too many machines to be returned in the call ? Should we also recommend a relook at filters defined in provider spec to list machines ?

Reply:
I think it could be possible in a provider with 1000s or millions of VM managed and accessible by the `provider-credentials` given here. However, only a handful of them (~10) might be relevant to the supplied `provider-spec`.

 I shall add a sentence about filtering the instances in the above call.

Reply:
Added a line on it. - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR735